File: expressjs_libraryapi-main/src/services/DataAccessor.ts
export type DataAccessor<Item, Create, Update>={
    /**
     * @throws {UnknownError}
     */
    create(book:Create):Promise<number>

    /**
     * @throws {NotFoundError}
     */
    readById(id:number):Promise<Item>

    /**
     * @throws {UnknownError}
     */
    read(page:number, perPage:number, query?:string):Promise<[number, Item[]]>

    /**
     * @throws {NotFoundError}
     */
    update(id:number, element:Update):Promise<void>

    /**
     * @throws {NotFoundError}
     */
    delete(id:number):Promise<void>
}

File: gin_libraryapi-main/services/dataaccessor.go
package services

type DataAccessor[Item any, Create any, Update any] interface {
	Create(element Create) (uint, error)
	Read(page uint, perPage uint, query string) (uint, *[]Item, error)
	ReadById(id uint) (*Item, error)
	Update(id uint, element Update) (bool, error)
	Delete(id uint) (bool, error)
}
